File: archive/llm_written_v1/gpt-5/gemm/go/gemm_parallel.go
package llm_written

import (
	"fmt"
	"runtime"
)

// Matrix is a rectangular 2D slice of float64.
type Matrix [][]float64

// getSize returns (rows, cols) of a rectangular matrix.
func getSize(m Matrix) (int, int) { return len(m), len(m[0]) }

// validateMatrix checks that the matrix is non-empty, rectangular, and numeric (float64 by type).
func validateMatrix(m Matrix, name string) error {
	if m == nil {
		return fmt.Errorf("%s must be a non-empty list of rows", name)
	}
	if len(m) == 0 {
		return fmt.Errorf("%s must have at least one row", name)
	}
	if len(m[0]) == 0 {
		return fmt.Errorf("%s must have at least one column", name)
	}
	cols := len(m[0])
	for r, row := range m {
		if row == nil {
			return fmt.Errorf("%s row %d is nil", name, r)
		}
		if len(row) != cols {
			return fmt.Errorf("%s is ragged: row %d has %d cols, expected %d", name, r, len(row), cols)
		}
	}
	return nil
}

// generateMatrix returns an rows x cols zero matrix.
func generateMatrix(rows, cols int) Matrix {
	out := make(Matrix, rows)
	for i := range out {
		out[i] = make([]float64, cols)
	}
	return out
}

// partialMatmul computes, for the provided packed blocks:
// C[m0+i_off][n0+j_off] += alpha * sum_{k=0..kb-1} A[i_off][k] * B[j_off][k]
func partialMatmul(A, B, C Matrix, alpha float64, m0, n0, kb int) {
	for iOff, Aik := range A {
		Ci := C[m0+iOff]
		for jOff, Bjk := range B {
			s := 0.0
			for k := 0; k < kb; k++ {
				s += Aik[k] * Bjk[k]
			}
			Ci[n0+jOff] += alpha * s
		}
	}
}

// transpose returns the transpose of a rectangular matrix.
func transpose(m Matrix) Matrix {
	r, c := getSize(m)
	out := make(Matrix, c)
	for j := 0; j < c; j++ {
		out[j] = make([]float64, r)
		for i := 0; i < r; i++ {
			out[j][i] = m[i][j]
		}
	}
	return out
}

// packMatrix returns a view (sub-slices) of matrix[b0:b1], each row sliced as row[a0:a1].
// It does not copy row data (reads only).
func packMatrix(matrix Matrix, a0, a1, b0, b1 int) Matrix {
	out := make(Matrix, b1-b0)
	for i := range out {
		out[i] = matrix[b0+i][a0:a1]
	}
	return out
}

const (
	defaultMB = 64
	defaultNB = 64
	defaultKB = 64
)

// Gemm computes C := alpha * A * B + beta * C.
// A is (m x k), B is (k x n), C is (m x n). If C == nil, a zero matrix is used.
// If MB/NB/KB are 0, sensible defaults are used.
func Gemm(A, B Matrix, alpha float64, C Matrix, beta float64, MB, NB, KB int) (Matrix, error) {
	if err := validateMatrix(A, "A"); err != nil {
		return nil, err
	}
	if err := validateMatrix(B, "B"); err != nil {
		return nil, err
	}

	m, k := getSize(A)
	k2, n := getSize(B)
	if k != k2 {
		return nil, fmt.Errorf("shape mismatch: A is (%d,%d), B is (%d,%d)", m, k, k2, n)
	}

	if C != nil {
		if err := validateMatrix(C, "C"); err != nil {
			return nil, err
		}
		m2, n2 := getSize(C)
		if m2 != m || n2 != n {
			return nil, fmt.Errorf("C has shape (%d,%d); expected (%d,%d)", m2, n2, m, n)
		}
		if beta != 1.0 {
			for i := 0; i < m; i++ {
				row := C[i]
				for j := 0; j < n; j++ {
					row[j] *= beta
				}
			}
		}
	} else {
		C = generateMatrix(m, n)
	}

	if alpha == 0.0 {
		return C, nil
	}

	// Default block sizes if not provided
	if MB == 0 {
		MB = defaultMB
	}
	if NB == 0 {
		NB = defaultNB
	}
	if KB == 0 {
		KB = defaultKB
	}

	// Small-input fast path: keep sequential to avoid overhead.
	const smallThreshold = 10 // roughly when parallel overhead outweighs benefit
	// const smallThreshold = 1_000 // roughly when parallel overhead outweighs benefit
	if m*n <= smallThreshold || m == 1 || n == 1 || k == 1 {
		return gemmSequential(A, B, alpha, C, 1.0 /*C already scaled*/, MB, NB, KB)
	}

	Bt := transpose(B)

	for n0 := 0; n0 < n; n0 += NB {
		n1 := n0 + NB
		if n1 > n {
			n1 = n
		}
		for k0 := 0; k0 < k; k0 += KB {
			k1 := k0 + KB
			if k1 > k {
				k1 = k
			}
			Bpack := packMatrix(Bt, k0, k1, n0, n1) // (nChunk x kChunk)

			// Parallelize the m0-loop with a bounded worker pool.
			mJobs := make(chan [2]int)
			workers := runtime.GOMAXPROCS(0)
			// Bound by number of chunks to avoid idle goroutines
			mChunks := (m + MB - 1) / MB
			if workers > mChunks {
				workers = mChunks
			}
			if workers < 1 {
				workers = 1
			}

			kb := k1 - k0

			done := make(chan struct{})
			for w := 0; w < workers; w++ {
				go func() {
					for mm := range mJobs {
						m0Local := mm[0]
						m1Local := mm[1]
						Apack := packMatrix(A, k0, k1, m0Local, m1Local) // (mChunk x kChunk)
						partialMatmul(Apack, Bpack, C, alpha, m0Local, n0, kb)
					}
					done <- struct{}{}
				}()
			}

			for m0 := 0; m0 < m; m0 += MB {
				m1 := m0 + MB
				if m1 > m {
					m1 = m
				}
				mJobs <- [2]int{m0, m1}
			}
			close(mJobs)
			for w := 0; w < workers; w++ {
				<-done
			}
		}
	}

	return C, nil
}

// gemmSequential is a faithful copy of the original sequential algorithm used for
// small inputs and for baseline testing. It assumes C has already been scaled by beta.
func gemmSequential(A, B Matrix, alpha float64, C Matrix, betaAlreadyApplied float64, MB, NB, KB int) (Matrix, error) {
	_ = betaAlreadyApplied // retained for API symmetry

	m, k := getSize(A)
	_, n := getSize(B)

	// Default block sizes if not provided
	if MB == 0 {
		MB = defaultMB
	}
	if NB == 0 {
		NB = defaultNB
	}
	if KB == 0 {
		KB = defaultKB
	}

	Bt := transpose(B)

	for n0 := 0; n0 < n; n0 += NB {
		n1 := n0 + NB
		if n1 > n {
			n1 = n
		}
		for k0 := 0; k0 < k; k0 += KB {
			k1 := k0 + KB
			if k1 > k {
				k1 = k
			}
			Bpack := packMatrix(Bt, k0, k1, n0, n1) // (nChunk x kChunk)

			for m0 := 0; m0 < m; m0 += MB {
				m1 := m0 + MB
				if m1 > m {
					m1 = m
				}
				Apack := packMatrix(A, k0, k1, m0, m1) // (mChunk x kChunk)

				kb := k1 - k0
				partialMatmul(Apack, Bpack, C, alpha, m0, n0, kb)
			}
		}
	}

	return C, nil
}

File: archive/llm_written_v1/gpt-5/gemm/python/gemm_parallel.py
from typing import List, Tuple, Optional
import numbers
import os
from concurrent.futures import ProcessPoolExecutor, as_completed

# Keep the public API (function names, signatures) the same as the sequential version


def get_size(matrix: List[List[float]]) -> Tuple[int, int]:
    """Return (rows, cols) of a rectangular matrix."""
    return len(matrix), len(matrix[0])


def validate_matrix(matrix: List[List[float]], name: str = "Matrix"):
    """Validate matrix is a non-empty rectangular list-of-lists of numbers."""
    if matrix is None or not isinstance(matrix, list):
        raise ValueError(f"{name} must be a non-empty list of rows.")
    if len(matrix) == 0:
        raise ValueError(f"{name}  must have at least one row.")
    if not isinstance(matrix[0], list) or len(matrix[0]) == 0:
        raise ValueError(f"{name} must have at least one column.")

    cols = len(matrix[0])
    for r, row in enumerate(matrix):
        if not isinstance(row, list):
            raise ValueError(f"{name} row {r} is not a list.")
        if len(row) != cols:
            raise ValueError(
                f"{name} is ragged: row {r} has {len(row)} cols, expected {cols}."
            )
        for c, v in enumerate(row):
            if not isinstance(v, numbers.Real):
                raise ValueError(f"{name}[{r}][{c}] is not a real number: {v!r}")


def generate_matrix(rows: int, cols: int) -> List[List[float]]:
    return [[0.0] * cols for _ in range(rows)]


def transpose(matrix: List[List[float]]) -> List[List[float]]:
    return [list(col) for col in zip(*matrix)]


def pack_matrix(
    matrix: List[List[float]], a0: int, a1: int, b0: int, b1: int
) -> List[List[float]]:
    return [row[a0:a1] for row in matrix[b0:b1]]


# Worker: compute contributions for a single (m-block x n-block) at a given k-slab
# Returns a dense block with alpha-scaled contributions for this slab only

def _compute_block_contrib(
    Apack: List[List[float]], Bpack: List[List[float]], kb: int, alpha: float
) -> List[List[float]]:
    # Apack shape: (mb x kb), Bpack shape: (nb x kb), both contiguous lists
    mb = len(Apack)
    nb = len(Bpack)
    # initialize result block (mb x nb) with zeros
    block = [[0.0] * nb for _ in range(mb)]

    # For each i in mb and j in nb, compute s = dot(Apack[i][:kb], Bpack[j][:kb])
    # Accumulate deterministically in the same k-order as baseline
    for i_off, Aik in enumerate(Apack):
        row_out = block[i_off]
        for j_off, Bjk in enumerate(Bpack):
            s = 0.0
            # fixed left-to-right summation order over k
            for k in range(kb):
                s += Aik[k] * Bjk[k]
            row_out[j_off] = alpha * s
    return block


def gemm(
    A: List[List[float]],
    B: List[List[float]],
    alpha: float = 1.0,
    C: Optional[List[List[float]]] = None,
    beta: float = 0,
    MB: int = 64,
    NB: int = 64,
    KB: int = 64,
):
    """
    Compute C := alpha * A * B + beta * C.

    A is (m x k), B is (k x n), C is (m x n). If C is None, a zero matrix is used.
    """
    from llm_written.python_openai.gemm.gemm_seq import gemm as seq_gemm  # for small-input fast path

    validate_matrix(A, "A")
    validate_matrix(B, "B")

    m, k = get_size(A)
    k2, n = get_size(B)

    if k != k2:
        raise ValueError(f"Shape mismatch: A is ({m},{k}), B is ({k2},{n}).")

    # Prepare C and scale by beta if needed (same semantics as baseline)
    if C is not None:
        validate_matrix(C, "C")
        m2, n2 = get_size(C)
        if m2 != m or n2 != n:
            raise ValueError(f"C has shape ({m2},{n2}); expected ({m},{n}).")
        if beta != 1.0:
            for i in range(m):
                row = C[i]
                for j in range(n):
                    row[j] *= beta
    else:
        C = generate_matrix(m, n)

    if alpha == 0.0:
        return C

    # Small-input or single-core fast path: defer to sequential reference
    total_elems = m * n
    cpu_cnt = os.cpu_count() or 1
    # Heuristic threshold: below this, the overhead of processes dominates
    if total_elems <= 1024 or cpu_cnt <= 1:
        return seq_gemm(A, B, alpha=alpha, C=C, beta=1.0, MB=MB, NB=NB, KB=KB)
        # beta is already applied above, so pass beta=1.0 to avoid rescaling

    Bt = transpose(B)

    # Set up a reusable process pool bounded by CPU and by 4 to avoid oversubscription
    max_workers_global = min(cpu_cnt, max(1, (m + MB - 1) // MB), 4)

    executor: Optional[ProcessPoolExecutor] = None
    try:
        if max_workers_global > 1:
            executor = ProcessPoolExecutor(max_workers=max_workers_global)
    except Exception:
        executor = None  # Fallback to sequential in restricted environments

    try:
        # Outer loops over tiles in deterministic order matching baseline
        for n0 in range(0, n, NB):
            n1 = min(n0 + NB, n)
            for k0 in range(0, k, KB):
                k1 = min(k0 + KB, k)
                kb = k1 - k0
                # Pack B for this (n0, k0)
                Bpack = pack_matrix(Bt, k0, k1, n0, n1)

                # Create tasks across m-tiles for this (n0, k0)
                m_tiles = []
                tasks = []
                for m0 in range(0, m, MB):
                    m1 = min(m0 + MB, m)
                    Apack = pack_matrix(A, k0, k1, m0, m1)
                    m_tiles.append((m0, m1))
                    tasks.append((m0, Apack))

                blocks_by_m0 = {}
                if executor is None:
                    # Sequential fallback
                    for m0, Apack in tasks:
                        block = _compute_block_contrib(Apack, Bpack, kb, alpha)
                        blocks_by_m0[m0] = block
                else:
                    future_map = {}
                    try:
                        for (m0, Apack) in tasks:
                            fut = executor.submit(
                                _compute_block_contrib, Apack, Bpack, kb, alpha
                            )
                            future_map[fut] = m0
                    except Exception:
                        # Submission failure => fallback to sequential for this slab
                        executor = None
                        blocks_by_m0.clear()
                        for m0, Apack in tasks:
                            block = _compute_block_contrib(Apack, Bpack, kb, alpha)
                            blocks_by_m0[m0] = block
                    else:
                        for f in as_completed(future_map):
                            m0_key = future_map[f]
                            block = f.result()
                            blocks_by_m0[m0_key] = block

                # Merge results into C in deterministic m0 order
                for m0, m1 in m_tiles:
                    block = blocks_by_m0[m0]
                    mb = m1 - m0
                    nb = n1 - n0
                    for i_off in range(mb):
                        Ci = C[m0 + i_off]
                        brow = block[i_off]
                        for j_off in range(nb):
                            Ci[n0 + j_off] += brow[j_off]
    finally:
        if executor is not None:
            executor.shutdown()

    return C
